subsys: lorawan: add devicetime request support

RomainPelletant opened the pull request:
Add devicetime request support:
-> Function to request devicetime : optionnally force sending message/optionnally blocking call
-> Function to retrieve current time from loramac-node stack
Fixes #55072

Commit:
lorawan: change downlink callback

Change downlink callback to transport data using flags.
Related to Issue #55072

Signed-off-by: romain pelletant <romainp@kickmaker.net>

teburd commented:
@JordanYates please take a look, thanks!

JordanYates commented:
How does this relate to the LoRaWAN clock synchronisation added in #51328?

RomainPelletant commented:
@JordanYates This PR is related to MAC DeviceTime commands (DeviceTimeReq, DeviceTimeAns) supported in LoRaWAN 1.0.3.

<img width="535" alt="Capture d’écran 2023-03-08 à 21 51 13" src="https://user-images.githubusercontent.com/91246595/223846969-debca447-3198-4580-b9b2-de9373ac59f4.png">

JordanYates commented on subsys/lorawan/lorawan.c:
```diff
@@ -359,6 +374,46 @@ int lorawan_set_region(enum lorawan_region region)
+	}
+
+	if (sync_call) {
+		k_sem_reset(&mlme_confirm_sem);
```
Arbitrarily resetting this semaphore seems dangerous. What happens if this another action is already running from another thread?

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -318,6 +318,25 @@ void lorawan_get_payload_sizes(uint8_t *max_next_payload_size,
+int lorawan_request_time(bool force_request, bool sync_call);
+
+/**
+ * @brief Read current time from LoRaWAN stack
```
Comment should probably make a note that the time returned is seconds from the GPS epoch

RomainPelletant commented:
Modifications will be done in few days

RomainPelletant commented:
Updated according to comments

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -166,10 +166,12 @@ struct lorawan_downlink_cb {
 		   int16_t rssi, int8_t snr,
-		   uint8_t len, const uint8_t *data);
+		   uint8_t len, const uint8_t *data,
+		   bool time);
```
Given we are breaking API compatibility here anyway, my preference would be to change `bool data_pending` to `uint8_t  flags`, together with a definition like:
```
enum lorwawan_dl_flags {
  LORAWAN_DATA_PENDING = BIT(0),
  LORAWAN_TIME_UPDATED = BIT(1),
};
```
This would let us add additional booleans (OTA update related?) without continuously breaking the compatibility.

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -330,6 +332,24 @@ void lorawan_get_payload_sizes(uint8_t *max_next_payload_size,
+/**
+ * @brief Read current time from LoRaWAN stack
+ *
+ * @return current time in unix epoch format
```
What does this function return if no time request has executed?

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -330,6 +332,24 @@ void lorawan_get_payload_sizes(uint8_t *max_next_payload_size,
+ * Append MAC DevTimeReq command. It will be processed on next send
+ * message or force sending empty message to request time immediately.
+ *
+ * @param force_request Force sending an empty message execute request
```
The following seems clearer to me:
`@param force_request Immediately send an empty message to execute the request`

JordanYates commented on subsys/lorawan/lorawan.c:
```diff
@@ -159,7 +159,8 @@ static void mcps_indication_handler(McpsIndication_t *mcps_indication)
 			       mcps_indication->BufferSize,
-			       mcps_indication->Buffer);
+			       mcps_indication->Buffer,
+				   mcps_indication->DeviceTimeAnsReceived);
```
Not aligned with the rest of the function.

JordanYates reviewed (changes requested):
Updating the callback signature should be a separate commit from the `DeviceTimeReq` functionality.

Commit:
lorawan: add devicetime request support

Add devicetime request support.
Update lorawan sample
Related to Issue #55072

Signed-off-by: romain pelletant <romain.pelletant@fullfreqs.com>

RomainPelletant commented:
Updated according to comments.
Split in two commits for callback signature update and the second for device time request feature.

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -161,13 +168,13 @@ struct lorawan_downlink_cb {
 	 *
 	 * @param port Port message was sent on
-	 * @param data_pending Network server has more downlink packets pending
+	 * @param flags Received flags (data pending)
```
Downlink data flags (see @ref lorawan_dl_flags)

JordanYates commented on samples/subsys/lorawan/class_a/src/main.c:
```diff
@@ -31,8 +31,9 @@ static void dl_callback(uint8_t port, uint8_t flags,
 {
-	LOG_INF("Port %d, Pending %d, RSSI %ddB, SNR %ddBm",
-			port, flags & LORAWAN_DATA_PENDING, rssi, snr);
+	LOG_INF("Port %d, Pending %d, RSSI %ddB, SNR %ddBm Time %d",
```
Missing comma before "Time"

JordanYates commented on samples/subsys/lorawan/class_a/src/main.c:
```diff
@@ -31,8 +31,9 @@ static void dl_callback(uint8_t port, uint8_t flags,
-			port, flags & LORAWAN_DATA_PENDING, rssi, snr);
+	LOG_INF("Port %d, Pending %d, RSSI %ddB, SNR %ddBm Time %d",
+			port, flags & LORAWAN_DATA_PENDING, rssi, snr,
+			flags & LORAWAN_TIME_UPDATED);
```
Flags prints should probably be done with `!!(flags & LORAWAN_TIME_UPDATED)`, otherwise the text will be `0` or `2`

JordanYates reviewed (changes requested):
Looking better, still the pending question about what `lorawan_get_time` returns if no time has been received.

RomainPelletant commented:
Requested changes applied.
Currently, as  _lorawan_get_time_ could return an inconsistent time (begins from epoch 0) and requires user to check if date is valid, a better approach could be based on _lorawan_clock_sync_get(uint32_t *gps_time)_

1. Declare static boolean variable in lorawan.c
2. On DevTimeAnswer received, toggle boolean value to true
3. _lorawan_get_time_ returns -EAGAIN while DevTime has been updated

The new prototype would be:
```
/**
 * @brief Retrieve the current time from LoRaWAN stack
 *
 * The epoch started on 1970-01-01T00:00:00Z.
 *
 * @param epoch_time Synchronized time in epoch format truncated to 32-bit.
 *
 * @return 0 if successful, -EAGAIN if the time is not yet updated.
 */
int lorawan_get_time(uint32_t *epoch_time);
```

This avoid user to check time validity

JordanYates commented:
Your proposed changes make sense to me. I think it just needs to be clear the difference between `lorawan_get_time` and `lorawan_clock_sync_get`

Commit:
lorawan: replace booleans by atomic flags

ADR status and devtime updated flags merged into an atomic bits array.
Related to Issue #55072

Signed-off-by: romain pelletant <romain.pelletant@fullfreqs.com>

RomainPelletant commented:
@JordanYates Please let me know if you think I should add more information about devtimerequest (lorawan_request_device_time) as details doxygen part and maybe add apptimereq (lorawan_clock_sync_get) informations in the same way

martinjaeger commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -330,6 +338,29 @@ void lorawan_get_payload_sizes(uint8_t *max_next_payload_size,
+ */
+int lorawan_request_device_time(bool force_request);
+
+/**
+ * @brief Retrieve the current time from LoRaWAN stack updated by
+ * DeviceTimeAns on MAC layer.
+ *
+ * The epoch starts at 1970-01-01T00:00:00Z.
+ *
+ * @param epoch_time Synchronized time in epoch format truncated to 32-bit.
+ *
+ * @return 0 if successful, -EAGAIN if the time is not yet updated.
+ */
+int lorawan_device_time_get(uint32_t *epoch_time);
```
I suggest to mention clearly that this is UNIX epoch and not GPS epoch (which is used in AppClockSync):

```suggestion
/**
 * @brief Retrieve the current UNIX timestamp from LoRaWAN stack updated by
 * DeviceTimeAns on MAC layer.
 *
 * The UNIX epoch starts at 1970-01-01T00:00:00Z.
 *
 * @param unix_time Synchronized time in seconds truncated to 32-bit.
 *
 * @return 0 if successful, -EAGAIN if the time is not yet updated.
 */
int lorawan_device_time_get(uint32_t *unix_time);
```

JordanYates replied:
Are you sure about that? As shown in an earlier screenshot, section 5.9 of the spec says:

> (*) The GPS epoch (i.e., January 6, 1980 00:00:00 UTC) is used as origin. The seconds field is the number of seconds elapsed since the origin. This field increases monotonically by 1 every second. To convert this field to UTC time, the leap seconds SHALL be taken into account

martinjaeger replied:
The timestamp is transferred in GPS epoch format, but this function gets the time through [SysTimeGet](https://github.com/zephyrproject-rtos/loramac-node/blob/842413c5fb98707eb5f26e619e8e792453877897/src/system/systime.h#L122), which uses the UNIX epoch.

See here for the code section doing the conversion: https://github.com/zephyrproject-rtos/loramac-node/blob/842413c5fb98707eb5f26e619e8e792453877897/src/mac/LoRaMac.c#L2552-L2578

However, it looks like it is ignoring the leap seconds...

JordanYates replied:
Great, so the time is only out by 20 something seconds. I would advocate for translating the value back to GPS time in our function then. Let the user perform conversions starting from a correct time, together with any knowledge of leap seconds they may have.

JordanYates replied:
> Now that Zephyr has a RTC API we could use both DeviceTimeReq and AppClockSync to set the time of the RTC instead of coming up with custom clock/time APIs based on uncommon epochs. Thoughts?

Not really a fan, are we adding `gmtime` into our libraries now? It makes more sense to me to leave it to the user to perform some action. Even using the RTC API would require providing the device, so it couldn't be automatic either.

martinjaeger commented on samples/subsys/lorawan/class_a/src/main.c:
```diff
@@ -27,11 +27,13 @@ LOG_MODULE_REGISTER(lorawan_class_a);
 {
-	LOG_INF("Port %d, Pending %d, RSSI %ddB, SNR %ddBm", port, data_pending, rssi, snr);
+	LOG_INF("Port %d, Pending %d, RSSI %ddB, SNR %ddBm, Time %d",
+			port, flags & LORAWAN_DATA_PENDING, rssi, snr,
```
For consistency, use `!!(...)` also here:
```suggestion
			port, !!(flags & LORAWAN_DATA_PENDING), rssi, snr,
```

martinjaeger reviewed (changes requested):
Hi @RomainPelletant,

Thanks for the PR, which is coming at the right time. I'm currently working on LoRaWAN Class B implementation, which needs the DeviceTimeReq as well.

The switch to atomic flags instead of booleans doesn't seem to be related to the PR, right? I'm not against it, but would like to understand the reason.

Only really got some minor comments, otherwise it's ready to go from my POV.

Also Zephyr v3.5.0 has just been released, so the merge window is open for new features.

JordanYates commented:
> @martinjaeger I switched to atomic to treat boolean ADR and boolean time updated at least once into a single variable. I can revert it if you prefer.

Keep the single flags variable instead of multiple `bool` please. As mentioned previously, either option is an API change, but the flags lets us continue to extend the options without further breaks.

> Beside, I can switch to GPS epoch time if you prefer too : let me test it again on a gateway on next Wednesday.

Please do this, it is preferable over a Unix time which we know is significantly wrong.
If possible, try and return the subseconds portion as well, I suspect it will be useful for class B as well as enabling generally better time synchronisation.

martinjaeger commented:
I agree with @JordanYates. The flags are fine and I understand the reason behind them now. Also agree to use GPS epoch and leave conversion to proper unix timestamp to the user.

RomainPelletant commented:
You are right with leap seconds. When LoRaWAN DevTimeAnswer gives 2023-10-24T10:06:14, the current (real) time is 2023-10-24T10:05:58 (with human reading tolerance :))

RomainPelletant commented:
@martinjaeger @JordanYates GPS based time is missing at this time but I would test it before pushing something

martinjaeger commented:
@RomainPelletant do you have any updates? I'd be happy to support with testing.

RomainPelletant commented:
@martinjaeger Unfortunately, I don't have the hardware anymore but should get it back in the end of this week

RomainPelletant commented:
WIP

shilga commented:
@martinjaeger @RomainPelletant 
what is the current state? Can we somehow move forward in implementing `DeviceTimeReq` MAC command? I'm happy to support or do what is left.

martinjaeger commented:
@shilga I don't need this feature anymore myself anymore, but I'm happy to review if you pick up this PR. There is not much missing, see the review comments for details.

shilga commented:
> @shilga There is not much missing, see the review comments for details.

That is what is unclear to me, what is still missing. It's mainly the GPS / Unix timestamp problem?
